rust/src/sign.rs

```diff
@@ -1,6 +1,235 @@
 use crate::{base64::ToUrlSafeBase64, http::HeaderName, jws::JwsHeader, openssl, Error};
 use indexmap::IndexMap;
-use std::fmt;
+use std::{fmt, future::Future};
+
+pub struct Unset;
+
+#[derive(Default)]
+pub struct SignerBuilder<'a, K, Pk, B, M, P> {
+    kid: K,
+    private_key: Pk,
+    body: B,
+    method: M,
+    path: P,
+    headers: IndexMap<HeaderName<'a>, &'a [u8]>,
+    jws_jku: Option<&'a str>,
+}
+
+impl<K, Pk, B, M, P> fmt::Debug for SignerBuilder<'_, K, Pk, B, M, P> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "Signer")
+    }
+}
+
+impl<'a> SignerBuilder<'a, Unset, Unset, Unset, Unset, Unset> {
+    pub fn new() -> Self {
+        SignerBuilder {
+            kid: Unset,
+            private_key: Unset,
+            body: Unset,
+            method: Unset,
+            path: Unset,
+            headers: <_>::default(),
+            jws_jku: <_>::default(),
+        }
+    }
+}
+
+impl<'a, Pk, B, M, P> SignerBuilder<'a, Unset, Pk, B, M, P> {
+    pub fn set_kid(self, kid: &str) -> SignerBuilder<'a, &str, Pk, B, M, P> {
```

Comment:
[nit]
Maybe let's go with just `kid` and `private_key` so the naming matches that of the other fields in `Signer` below.